### ISSUE TITLE
Should not use locale-dependent C standard-library functions

### ISSUE THREAD

**Issue by hsivonen** [mozilla/libprio#20]
libprio uses locale-dependent character classification and manipulation functions like, `isdigit`, `isxdigit`, `isupper`, `islower` and `toupper`.

This is problematic.

First, libprio passes `char` to these, but these are defined to take `int` whose value is the value of an `unsigned char` or `EOF`. On platforms where `char` is signed, passing it directly to these functions without casting to `unsigned char` first results in sign-extension and Undefined Behavior when the high bit of the `char` is set.

Second, these are locale-dependent, which makes them inappropriate for processing protocol text. Even the ones that aren't locale-dependent by standard are made so by Microsoft.

See:
https://stackoverflow.com/questions/2898228/can-isdigit-legitimately-be-locale-dependent-in-c
https://daniel.haxx.se/blog/2018/01/30/isalnum-is-not-my-friend/
https://daniel.haxx.se/blog/2008/10/15/strcasecmp-in-turkish/

I suggest doing the intended comparisons with `char` literals. That is, `('0' <= c && c <= '9')` in place of `isdigit(c)`.

**Comment by henrycg** [comment]
Thanks for pointing this out! Your proposed fix sounds good to me. Would you be wiling to submit a pull request?

**Comment by rhelmer** [comment]
> libprio uses locale-dependent character classification and manipulation functions like, isdigit, isxdigit, isupper, islower and toupper.

Just for clarity I took a quick look and some of these usages are in msgpack (this uses `std::toupper`) and the copy of NSS MPI in libprio... this is a bit confusing and needs some explanation.

The `./mpi` dir is actually part of NSS and lives in mozilla-central, but it's not included as part of libnss, which the standalone libprio uses. This is only used when building libprio completely standalone.

So, the way we resolved this for Firefox is that we remove the `./mpi` dir when we vendor libprio, and build against the in-tree mozilla-central copy instead.

Therefore, excluding these I only see these in Prio proper:

`isxdigit`: https://github.com/mozilla/libprio/blob/bfbcf1f95001047c61f4a29c3e68499b2f0c9f3b/prio/encrypt.c#L114
`toupper`: https://github.com/mozilla/libprio/blob/master/prio/encrypt.c#L41

Most of these are in NSS MPI and should be fixed in mozilla-central. The one use in msgpack we should consider contributing to the upstream repo in https://github.com/msgpack/msgpack-c/
